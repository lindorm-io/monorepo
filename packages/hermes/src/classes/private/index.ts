export * from "./HermesMetadata";
export * from "./HermesRegistry";
export * from "./HermesScanner";
