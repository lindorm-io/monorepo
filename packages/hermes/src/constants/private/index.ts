export * from "./checksum-store";
export * from "./domain";
export * from "./event-store";
export * from "./saga-store";
export * from "./view-store";
