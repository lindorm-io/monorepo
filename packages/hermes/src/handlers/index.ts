export * from "./HermesAggregateCommandHandler";
export * from "./HermesAggregateErrorHandler";
export * from "./HermesAggregateEventHandler";
export * from "./HermesSagaErrorHandler";
export * from "./HermesSagaEventHandler";
export * from "./HermesSagaIdHandler";
export * from "./HermesSagaTimeoutHandler";
export * from "./HermesViewErrorHandler";
export * from "./HermesViewEventHandler";
export * from "./HermesViewIdHandler";
export * from "./HermesViewQueryHandler";
