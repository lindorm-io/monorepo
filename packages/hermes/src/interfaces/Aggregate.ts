import { ClassLike, Dict } from "@lindorm/types";
import { AggregateData } from "../types";
import { IHermesMessage } from "./HermesMessage";

export interface IAggregate<S extends Dict = Dict> extends AggregateData {
  apply(causation: IHermesMessage, event: ClassLike): Promise<void>;
  load(event: IHermesMessage): void;
  toJSON(): AggregateData<S>;
}
