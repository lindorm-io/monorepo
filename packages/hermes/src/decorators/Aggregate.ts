import { Constructor } from "@lindorm/types";
import { AggregateDecoratorOptions } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function Aggregate(options: AggregateDecoratorOptions = {}): ClassDecorator {
  return function (target) {
    const { name } = extractNameData(target.name);
    globalHermesMetadata.addAggregate({
      encryption: options.encryption || false,
      name: options.name || name,
      namespace: options.namespace || null,
      target: target as unknown as Constructor,
    });
  };
}
