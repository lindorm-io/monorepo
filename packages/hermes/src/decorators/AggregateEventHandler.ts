import { Constructor, Dict } from "@lindorm/types";
import { AggregateEventHandlerDescriptor } from "../types";
import { globalHermesMetadata } from "../utils/private";

export function AggregateEventHandler<C extends Constructor, S extends Dict>(
  EventClass: C,
): AggregateEventHandlerDescriptor<C, S> {
  return function (target, key, descriptor) {
    globalHermesMetadata.addHandler({
      conditions: null,
      decorator: "AggregateEventHandler",
      encryption: false,
      handler: descriptor.value,
      key: key.toString(),
      schema: null,
      target: target.constructor as Constructor,
      trigger: EventClass,
    });
  };
}
