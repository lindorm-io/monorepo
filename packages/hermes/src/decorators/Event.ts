import { Constructor } from "@lindorm/types";
import { EventDecoratorOptions } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function Event(options?: EventDecoratorOptions): ClassDecorator {
  return function (target) {
    const { name, version } = extractNameData(target.name);
    globalHermesMetadata.addEvent({
      name: options?.name || name,
      target: target as unknown as Constructor,
      version: options?.version || version,
    });
  };
}
