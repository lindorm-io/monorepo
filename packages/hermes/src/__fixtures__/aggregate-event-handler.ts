import { HermesAggregateEventHandler } from "../handlers";
import { AggregateEventHandlerOptions } from "../types";
import { TEST_AGGREGATE_OPTIONS } from "./aggregate";
import {
  TEST_HERMES_EVENT,
  TEST_HERMES_EVENT_CREATE,
  TEST_HERMES_EVENT_DESTROY,
  TEST_HERMES_EVENT_DESTROY_NEXT,
  TEST_HERMES_EVENT_MERGE_STATE,
  TEST_HERMES_EVENT_SET_STATE,
  TEST_HERMES_EVENT_THROWS,
} from "./hermes-event";

export const TEST_AGGREGATE_EVENT_HANDLER_OPTIONS: AggregateEventHandlerOptions = {
  aggregate: {
    name: TEST_AGGREGATE_OPTIONS.name,
    context: TEST_AGGREGATE_OPTIONS.context,
  },
  eventName: TEST_HERMES_EVENT.name,
  handler: jest.fn().mockImplementation(async () => {}),
};

export const TEST_AGGREGATE_EVENT_HANDLER = new HermesAggregateEventHandler(
  TEST_AGGREGATE_EVENT_HANDLER_OPTIONS,
);

export const TEST_AGGREGATE_EVENT_HANDLER_CREATE = new HermesAggregateEventHandler({
  ...TEST_AGGREGATE_EVENT_HANDLER_OPTIONS,
  eventName: TEST_HERMES_EVENT_CREATE.name,
  handler: jest.fn().mockImplementation(async (ctx) => {
    ctx.mergeState({ created: true });
  }),
});

export const TEST_AGGREGATE_EVENT_HANDLER_DESTROY = new HermesAggregateEventHandler({
  ...TEST_AGGREGATE_EVENT_HANDLER_OPTIONS,
  eventName: TEST_HERMES_EVENT_DESTROY.name,
  handler: jest.fn().mockImplementation(async (ctx) => {
    ctx.destroy();
  }),
});

export const TEST_AGGREGATE_EVENT_HANDLER_DESTROY_NEXT = new HermesAggregateEventHandler({
  ...TEST_AGGREGATE_EVENT_HANDLER_OPTIONS,
  eventName: TEST_HERMES_EVENT_DESTROY_NEXT.name,
  handler: jest.fn().mockImplementation(async (ctx) => {
    ctx.destroyNext();
  }),
});

export const TEST_AGGREGATE_EVENT_HANDLER_MERGE_STATE = new HermesAggregateEventHandler({
  ...TEST_AGGREGATE_EVENT_HANDLER_OPTIONS,
  eventName: TEST_HERMES_EVENT_MERGE_STATE.name,
  handler: jest.fn().mockImplementation(async (ctx) => {
    ctx.mergeState({ merge: ctx.event });
  }),
});

export const TEST_AGGREGATE_EVENT_HANDLER_SET_STATE = new HermesAggregateEventHandler({
  ...TEST_AGGREGATE_EVENT_HANDLER_OPTIONS,
  eventName: TEST_HERMES_EVENT_SET_STATE.name,
  handler: jest.fn().mockImplementation(async (ctx) => {
    ctx.mergeState({ set: "state" });
  }),
});

export const TEST_AGGREGATE_EVENT_HANDLER_THROWS = new HermesAggregateEventHandler({
  ...TEST_AGGREGATE_EVENT_HANDLER_OPTIONS,
  eventName: TEST_HERMES_EVENT_THROWS.name,
  handler: jest.fn().mockImplementation(async () => {
    throw new Error("throw");
  }),
});
