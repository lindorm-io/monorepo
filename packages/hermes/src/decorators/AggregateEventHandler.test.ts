import { Dict } from "@lindorm/types";
import { AggregateEventCtx } from "../types/handlers/aggregate-event-handler";
import { globalHermesMetadata } from "../utils/private";
import { Aggregate } from "./Aggregate";
import { AggregateEventHandler } from "./AggregateEventHandler";
import { Event } from "./Event";

describe("AggregateEventHandler Decorator", () => {
  test("should add metadata", () => {
    @Event()
    class TestEvent {}

    @Aggregate()
    class TestAggregateEventHandlerAggregate {
      @AggregateEventHandler(TestEvent)
      public async onTestAggregateDomainEvent(ctx: AggregateEventCtx<TestEvent, Dict>) {}
    }

    expect(
      globalHermesMetadata.getAggregate(TestAggregateEventHandlerAggregate),
    ).toMatchSnapshot();
  });
});
