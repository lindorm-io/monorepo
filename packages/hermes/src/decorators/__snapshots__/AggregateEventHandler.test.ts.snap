// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AggregateEventHandler Decorator should add metadata 1`] = `
{
  "encryption": false,
  "handlers": [
    {
      "conditions": null,
      "decorator": "AggregateEventHandler",
      "encryption": false,
      "handler": [Function],
      "key": "onTestAggregateDomainEvent",
      "schema": null,
      "target": [Function],
      "trigger": [Function],
    },
  ],
  "name": "test_aggregate_event_handler_aggregate",
  "namespace": null,
  "target": [Function],
}
`;
